runtime/lua/vim/shared.lua

```diff
@@ -267,18 +267,25 @@ function vim.tbl_deep_extend(behavior, ...)
 end
 
 --- Deep compare values for equality
+---
+--- Tables are compared recursively unless they both provide the `eq` methamethod.
+--- All other types are compared using the equality `==` operator.
+--@param a first value
+--@param b second value
+--@returns `true` if values are equals, else `false`.
 function vim.deep_equal(a, b)
   if a == b then return true end
   if type(a) ~= type(b) then return false end
   if type(a) == 'table' then
-    -- TODO improve this algorithm's performance.
+    local keyset = {}
     for k, v in pairs(a) do
       if not vim.deep_equal(v, b[k]) then
         return false
       end
+      keyset[k] = true
     end
-    for k, v in pairs(b) do
-      if not vim.deep_equal(v, a[k]) then
+    for k, _ in pairs(b) do
+      if not keyset[k] then
```

Comment:
Can't this just be `a[k] ~= nil` and eliminate `keyset` ?

Reply:
Yes that would be the same! Will change it.